Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -161,97 +163,82 @@ public void Serialize<T>(Stream stream, T item)
+                if (item == null)
+                {
+                    new IonNullSerializer().Serialize(writer, (object)null);
+                }
+                else if (item is bool)
+                {
+                    new IonBooleanSerializer().Serialize(writer, Convert.ToBoolean(item));
+                }
+                else if (item is int)
+                {
+                    new IonIntSerializer().Serialize(writer, Convert.ToInt32(item));
+                }
+                else if (item is long)
+                {
+                    new IonLongSerializer().Serialize(writer, Convert.ToInt64(item));
+                }
+                else if (item is float)
+                {
+                    new IonFloatSerializer().Serialize(writer, Convert.ToSingle(item));
+                }
+                else if (item is double)
+                {
+                    new IonDoubleSerializer().Serialize(writer, Convert.ToDouble(item));
+                }
+                else if (item is decimal)
+                {
+                    new IonDecimalSerializer().Serialize(writer, Convert.ToDecimal(item));
+                }
+                else if (item is BigDecimal)
+                {
+                    new IonBigDecimalSerializer().Serialize(writer, (BigDecimal)(object)item);
+                }
+                else if (item is byte[])
+                {
+                    new IonByteArraySerializer().Serialize(writer, (byte[])(object)item);
+                }
+                else if (item is string)
+                {
+                    new IonStringSerializer().Serialize(writer, item as string);
+                }
+                else if (item is SymbolToken)
+                {
+                    new IonSymbolSerializer().Serialize(writer, (SymbolToken)(object)item);
+                }
+                else if (item is DateTime)
+                {
+                    new IonDateTimeSerializer().Serialize(writer, (DateTime)(object)item);
+                }
+                else if (item is System.Collections.IList) 
+                {
+                    NewIonListSerializer(item.GetType()).Serialize(writer, (System.Collections.IList)(object)item);
+                }
+                else if (item is Guid) 
+                {
+                    new IonGuidSerializer(options).Serialize(writer, (Guid)(object)item);
+                }
+                else if (item is object) 
+                {
+                    new IonObjectSerializer(this, options, item.GetType()).Serialize(writer, item);
+                }
+                else
+                {
+                    throw new NotSupportedException($"{typeof(T)} is not supported for serialization");
+                }
             }
-
-            if (item is object) 
+            finally
             {
-                new IonObjectSerializer(this, options, item.GetType()).Serialize(writer, item);
-                return;
+                currentDepth--;

```

Comment:
The current depth should probably live on the context object that gets passed around. As mentioned above, this is not thread safe.

Reply:
Turns out the `IIonReader` tracks current depth so now we just check that instead of tracking it ourselves.